src/server/app-router.ts

```diff
@@ -635,27 +633,32 @@ function redirectToLogin(redirectOptions?: RedirectOptions) {
     }
 }
 
-export function getUrlEncodedRedirectPathForCurrentUrl(): string | undefined {
-    const url = getCurrentUrl()
-    if (!url) {
+export function getUrlEncodedRedirectPathForCurrentPath(): string | undefined {
+    const path = getCurrentPath()
+    if (!path) {
         return undefined
     }
 
-    try {
-        const urlObj = new URL(url)
-        return encodeURIComponent(urlObj.pathname + urlObj.search)
-    } catch (e) {
-        console.warn('Current URL is not a valid URL')
-        return undefined
-    }
+    return encodeURIComponent(path)
 }
 
 // It's really common to want to redirect back to the exact location you are on
-// Next.js unfortunately makes this pretty hard, as server components don't have access to the URL
-// The only good way to do this is to set up some middleware and pass the URL down from the middleware
+// Next.js unfortunately makes this pretty hard, as server components don't have access to the path
+// The only good way to do this is to set up some middleware and pass the path down from the middleware
 // Since we have the requirement that people set up middleware with us anyway, it's easy for us to expose
 // this functionality
+export function getCurrentPath(): string | undefined {
+    const path = headers().get(CUSTOM_HEADER_FOR_PATH)
+    if (!path) {
+        console.warn('Attempting to redirect to the current path, but we could not find the current path in the headers. Is the middleware set up?')
+        return undefined
+    } else {
+        return path
+    }
+}
+
 export function getCurrentUrl(): string | undefined {
```

Comment:
[nit] We could also leave a docstring for deprecated, so it will show up in peoples IDE's:

```
/**
 * @deprecated getCurrentUrl is deprecated in favor of getCurrentPath
 */
export function getCurrentUrl(): string | undefined {
```

Reply:
done, thanks!